.ci/lint_r_code.R

```diff
@@ -0,0 +1,64 @@
+
+library(argparse)
+library(lintr)
+
+parser <- argparse::ArgumentParser()
+parser$add_argument(
+    "--source-dir"
+    , type = "character"
+    , help = "Fully-qualified directory to search for R files"
+)
+args <- parser$parse_args()
+
+SOURCE_DIR <- args[["source_dir"]]
+
+FILES_TO_LINT <- list.files(
+    path = SOURCE_DIR
+    , pattern = "\\.r$"
+    , all.files = TRUE
+    , ignore.case = TRUE
+    , full.names = TRUE
+    , recursive = TRUE
+    , include.dirs = FALSE
+)
+
+LINTERS_TO_USE <- list(
```

Comment:
Can we use more linters? For instance, `camel_case_linter`, `spaces_left_parentheses_linter`, `T_and_F_symbol_linter`, etc.

https://github.com/jimhester/lintr#available-linters

Reply:
ah ok now I see why you were asking about spaces and parentheses. I don't want to introduce `camel_case_linter` but I think these make sense to add:

* `T_and_F_symbol_linter`
* `spaces_left_parentheses_linter`
* `paren_brace_linter` (to your questions about changing `(){` to `() {`)
* `spaces_inside_linter`
* `semicolon_terminator_linter`
* `equals_na_linter`
* `function_left_parentheses_linter`

I know Jim Hester has warned in the past that some of these can make linting take a long time (since, for example, `function_left_parentheses_linter` requires actually interpreting the source code). Let me try adding these and see how it goes! Thanks for the nudge.

Reply:
Ah, nice! I think you should reach the agreement about the list of linters with @Laurae2.

Reply:
ok @Laurae2 here is my proposed list, could you take a look at this list and https://github.com/jimhester/lintr#available-linters and let me know if you agree / disagree? 

Once you and I agree, I'll make whatever remaining changes are needed to get the code into compliance with the linters we pick. I don't want to touch any more lines yet until we agree on the list of linters.

When I call a linter "expensive" below, I mean "requires interpretation of the code and not (in my opinion) worth that cost". These can be expensive because R is a dynamically-typed, interpreted language where people can do some very..."creative"... things that are not possible in stricter languages. Linting with `lintr` can become quite costly to run especially for the more complex linters like `extraction_operator_linter`.

**Use**

* `closed_curly_linter`
* `equals_na_linter`
* `function_left_parentheses_linter`
* `implicit_integer_linter`
* `infix_spaces_linter`
* `line_length_linter` (set to 120 just to avoid super-long lines)
* `no_tab_linter`
* `open_curly_linter`
* `paren_brace_linter`
* `semicolon_terminator_linter`
* `seq_linter`
* `spaces_inside_linter`
* `spaces_left_parentheses_linter`
* `todo_comment_linter`
* `trailing_blank_lines_linter`
* `trailing_whitespace_linter`
* `T_and_F_symbol_linter`

**Don't Use (reason why)**

* `object_usage_linter` (redundant with R CMD CHECK)
* `absolute_path_linter` (expensive and error-prone. superfluous with redundant with CRAN checks + CI we'll run on multiple OSes)
* `nonportable_path_linter` (expensive and error-prone. superfluous with redundant with CRAN checks + CI we'll run on multiple OSes)
* `pipe_continuation_linter` (irrelevant, factored out pipes in #2334 
* `assignment_linter` (expensive)
* `camel_case_linter` (expensive, not worth the effort for a project of this maturity, inconsistent with our current style)
* `commas_linter` (counters our goal of comma-first)
* `commented_code_linter` (expensive)
* `cyclocomp_linter` (expensive)
* `extraction_operator_linter` (expensive and error-prone)
* `object_length_linter` (expensive)
* `object_name_linter`: (expensive, not worth the effort for a project of this maturity)
* `single_quotes_linter` (expensive)
* `undesirable_function_linter` (unnecessary)
* `undesirable_operator_linter` (expensive, also we use `<<-` in some valid but non-standard ways)
* `unneeded_concatenation_linter` (expensive)

Reply:
Even though `object_usage_linter` is redundant with R CMD CHECK, perhaps we should keep using it?

About `line_length_linter`, should we keep it at 120 or the standard "default" 80 character maximum per line, as it is the R "standard"?

Reply:
I think we should not use `object_usage_linter`. In the past on other projects, I've found that it is error-prone and relies on the naming of scripts. So for example functions in R are lazily evaluated and `R CMD CHECK` will not complain if you define `.do_somemthing()` in `a.R` and call it in `b.R`, but `object_usage_linter` will complain and say "undefined global variable '.do_something'".

For `line_length_linter`, in my opinion 80 is overly strict and can be counterproductive, e.g. leading to people using uninformative short variable names to stay under the limit. The common standard of 80 characters in a lot of communities is [a carryover from the earliest days of computing](https://softwareengineering.stackexchange.com/a/148729) that is not really necessary any more with modern IDEs and shells. In my experience, 100 or 120 is a better compromise between meaningful variable names and excessive complexity caused by complicated one-liners in code.

Reply:
ok @Laurae2 given discussion [here](https://github.com/microsoft/LightGBM/pull/2437#discussion_r329351521) about installation from source, here is my updated proposal

**Use**

Added #2477 to #2302 . So now @Laurae2 my proposed list of linters is:

* `closed_curly_linter`
* `infix_spaces_linter`
* `line_length_linter (120)`
* `no_tab_linter`
* `open_curly_linter`
* `spaces_inside_linter`
* `spaces_left_parentheses_linter`
* `trailing_blank_lines_linter`
* `trailing_whitespace_linter`

**Use once `lintr` does a new release**

*documented in #2477*

* `equals_na_linter`
* `function_left_parentheses_linter`
* `implicit_integer_linter`
* `:paren_brace_linter`
* `semicolon_terminator_linter`
* `seq_linter`
* `todo_comment_linter`
* `T_and_F_symbol_linter`

**Do Not Use**

*same as original post*

Reply:
Just want to say that I'm totally agree with @jameslamb about the 80-chars line.

Reply:
@Laurae2 could you take another look at this PR, particularly let me know if you agree with the list in [this comment](https://github.com/microsoft/LightGBM/pull/2437#discussion_r329358150)? If you agree, then I'll do a tiny bit more refactoring to satisfy all those linters, and this will be good to merge. Thanks!

Reply:
I just rebased this to latest `master` to catch up with changes over the last two weeks. It is ready for review.